extensions/answer_summarizers/models.py

```diff
@@ -43,42 +43,79 @@
+class _HashableAnswer(object):
+    """Wraps an arbitrarily-complex answer dict object into an object that can
+    be hashed into built-in collections.
     """
-    if isinstance(value, list):
-        # Avoid needlessly wrapping a single value in a tuple.
-        if len(value) == 1:
-            return _get_hashable_value(value[0])
-        return tuple([_get_hashable_value(elem) for elem in value])
-    elif isinstance(value, dict):
-        return _get_hashable_value(
-            sorted([
-                (_get_hashable_value(key), _get_hashable_value(val))
-                for (key, val) in value.iteritems()]))
-    else:
-        return value
-
-
-def _count_answers(answer_dicts_list):
-    """Count an input list of answer objects using collections.Counter. This
-    returns a list of pairs with the first element being an answer object and
-    the second being the number of times it shows up in the input list.
+
+    @classmethod
+    def _hashable(cls, value):
+        """This function returns a hashable version of the input value.
+
+        It converts the built-in collections into their hashable counterparts
+        {list: (tuple), set: (frozenset), dict: (sorted tuple of pairs)}.
+        Additionally, their elements are converted to hashable values through
+        recursive calls. All other values are assumed to already be hashable.
+
+        For single-element lists and sets, the single element is returned by
+        itself.
+        """
+        if isinstance(value, list):
+            # Can't be sure that all elements of lists are hashable, so apply
+            # logic recursively to each element.
+            if len(value) == 1:
+                # Single-elements can be returned by themselves.
+                return cls._hashable(value[0])
+            else:
+                return tuple(cls._hashable(e) for e in value)
+        elif isinstance(value, set):
+            # Set elements are always hashable; don't need to call recursively.
+            if len(value) == 1:
+                # Single-elements can be returned by themselves.
+                return next(iter(value))
+            else:
+                return frozenset(value)
+        elif isinstance(value, dict):
+            # Dict keys are always hashable. Can't be sure that values are
+            # hashable.
+            return cls._hashable(  # Reuse existing {list: tuple} logic.
+                sorted((k, cls._hashable(v)) for k, v in value.iteritems()))
+        else:
+            return value  # Any other type is assumed to already be hashable.
+
+    def __init__(self, answer_dict):
+        self.value = answer_dict
+        self._hashable_value = self._hashable(self.value['answer'])
+
+    def __hash__(self):
+        return hash(self._hashable_value)
+
+    def __eq__(self, other):
+        if isinstance(other, _HashableAnswer):
+            return self._hashable_value == other._hashable_value  # pylint: disable=protected-access
```

Comment:
Any reason not to just make the hash public?

Reply:
I kept it protected because nothing else uses it, but I don't have a strong preference about that so I'll opt for public since you pointed it out :)